### DISCUSSION TITLE
Is it possible to allow a shorter input to `from_dict_of_dicts`?

### DISCUSSION THREAD

**Discussion by erelsgl** [networkx/networkx#4682] [Ideas]
Currently, `networkx.convert.from_dict_of_dicts` requires an input such as:

    {0: {1: {"weight": 2}}}

for a single edge (0,1) with weight 2. When there are many edges, the string "weight" takes a lot of space. Is it possible to allow a shorter input such as:

    {0: {1: 2}}

where "2" is automatically interpreted as weight?

(Currently, this returns `TypeError: 'int' object is not iterable`).

**Comment by MridulS** [comment]
`from_dict_of_dicts` currently requires the input dict to have the the keywords `{0: {1: {"weight": 2, "another_weight": 3}}}` in the edge attribute dictionary. To save on space you could try something like
``` python
edges = [(0, 1, 1000), (0, 2, 10), (0, 3, -10)]
G  = nx.Graph() # or DiGraph or MultiGraph
G.add_weighted_edges_from(edges)
```
This is assuming you have a list of edges instead of a adjacency dictionary.

If you have an adjacency dictionary you can preprocess it and convert it to an edgelist.
``` python
edges_adj = {0: {1: 100, 2: 123, 3: 1 }, 1: {3: 4}}
edges = [(u, v, edges_adj[u][v]) for u in edges_adj for v in edges_adj[u]]

```

**Reply by MridulS** [reply]
NOTE: The graph object will still take up the same space as it will internally save the edge attributes with the keyword `weight`. It will just save you space will working with the edges outside of the networkx Graph object.
``` python
>>> G.edges.data()
EdgeDataView([(0, 1, {'weight': 1000}), (0, 2, {'weight': 10}), (0, 3, {'weight': -10})])
```

**Comment by dschult** [comment]
It may also be worth stating that Python doesn't actually store the string "weight" multiple times.  Each string created in CPython is stored in a central memory location and later uses of the same string simply point to that central location. So the edgedata dictionary uses enough memory to store a pointer to the string, but not the string itself.
